Add blog post about deprecating `feature = "cargo-clippy"` (#1253)
rust-lang/blog.rust-lang.org: posts/2024-02-16-Clippy-deprecating-feature-cargo-clippy.md

---
layout: post
title: "Clippy: Deprecating `feature = \"cargo-clippy\"`"
author: The Clippy Team
release: true
---

Since Clippy [`v0.0.97`] and before it was shipped with `rustup`, Clippy
implicitly added a `feature = "cargo-clippy"`[^1] when linting your code with
`cargo clippy`.

> [^1]: It's likely, that you didn't even know that Clippy implicitly sets this
>     feature. This is intentional, as we stopped advertising and documenting this
>     a long time ago.

Reviewer: Maybe clarify that it was not a Cargo feature, but a config that impersonnated Cargo features.
```suggestion
[^1]: It's likely, that you didn't even know that Clippy implicitly sets this
    config (which was not a Cargo feature). This is intentional, as we stopped advertising and documenting this
    a long time ago.
```

```rust
#[allow(clippy::lint_name)]
```

The implicit `feature = "cargo-clippy"` has only been kept for backwards
compatibility, but will now be deprecated.

## Alternative

As there is a rare [use case] for conditional compilation depending on Clippy,
we will provide an alternative. So in the future you will be able to use:

```rust
#[cfg(clippy)]
```

## Transitioning

Should you have instances of `feature = "cargo-clippy"` in your code base, you
will see a warning from the new Clippy lint
[`clippy::deprecated_clippy_cfg_attr`][pr-12292]. This lint can automatically fix
your code. So if you should see this lint triggering, just run:

```
cargo clippy --fix -- -Aclippy::all -Wclippy::deprecated_clippy_cfg_attr
```

This will fix all instances in your code.

In addition, check your `.cargo/config` file for:

```toml
[target.'cfg(feature = "cargo-clippy")']
rustflags = ["-Aclippy::..."]
```

This file you have to update yourself, by either changing it to `cfg(clippy)` or
taking this opportunity to transition to [setting lint levels in
`Cargo.toml`][cargo-lints] directly.

## Motivation for Deprecation

There's a plan to stabilize [checking conditional compilation at compile

> time][rfc-3013], aka `cargo check -Zcheck-cfg`. If we were to keep `feature =

Reviewer: We may want to mention the [Call for Testing](https://github.com/rust-lang/rfcs/pull/3013#issuecomment-1936648479).

Author: Good idea, added in https://github.com/rust-lang/blog.rust-lang.org/pull/1253/commits/f56b2c68d73cef8856123c1cebfebdffed77cbf0

---
layout: post
title: "Clippy: Deprecating `feature = \"cargo-clippy\"`"
author: The Clippy Team
release: true
---

Since Clippy [`v0.0.97`] and before it was shipped with `rustup`, Clippy
implicitly added a `feature = "cargo-clippy"` config[^1] when linting your code
with `cargo clippy`.

[^1]: It's likely, that you didn't even know that Clippy implicitly sets this
    config (which was not a Cargo feature). This is intentional, as we stopped
    advertising and documenting this a long time ago.

> Back in the day (2016) this was necessary to allow, warn or deny lints with

Reviewer 2: Making it a little more precise:
```suggestion
Back in the day (2016) this was necessary to allow, warn or deny Clippy lints using
```

attributes:

```rust
#[cfg_attr(feature = "cargo-clippy", allow(clippy_lint_name))]
```

Doing this hasn't been necessary for a long time. Today, Clippy users will set
lint levels with tool lint attributes using the `clippy::` prefix:

```rust
#[allow(clippy::lint_name)]
```

The implicit `feature = "cargo-clippy"` has only been kept for backwards
compatibility, but will now be deprecated.

## Alternative

As there is a rare [use case] for conditional compilation depending on Clippy,
we will provide an alternative. So in the future you will be able to use:

```rust
#[cfg(clippy)]
```

## Transitioning

Should you have instances of `feature = "cargo-clippy"` in your code base, you
will see a warning from the new Clippy lint
[`clippy::deprecated_clippy_cfg_attr`][pr-12292]. This lint can automatically fix
your code. So if you should see this lint triggering, just run:

```
cargo clippy --fix -- -Aclippy::all -Wclippy::deprecated_clippy_cfg_attr
```

This will fix all instances in your code.

In addition, check your `.cargo/config` file for:

```toml
[target.'cfg(feature = "cargo-clippy")']
rustflags = ["-Aclippy::..."]
```

> This file you have to update yourself, by either changing it to `cfg(clippy)` or

Reviewer 2: Let's make it clear that no action is needed if no such thing exists:
```suggestion
If you have this config, you will have to update it yourself, by either changing it to `cfg(clippy)` or
```

taking this opportunity to transition to [setting lint levels in
`Cargo.toml`][cargo-lints] directly.

## Motivation for Deprecation

Currently, there's a [call for testing], in order to stabilize [checking
conditional compilation at compile time][rfc-3013], aka `cargo check

> -Zcheck-cfg`. If we were to keep the `feature = "cargo-clippy"` config, users
> would start seeing a lot of warnings when running `cargo clippy`. To work around

Reviewer: Missed this one: the warnings would be on any `feature = "cargo-clippy"` conditions, for any Cargo command that invokes `rustc` (so check, build, doc, test, ...), it would not be restricted to `cargo clippy`.
```suggestion
-Zcheck-cfg`. If we were to keep the `feature = "cargo-clippy"` config, users
would start seeing a lot of warnings on their `feature = "cargo-clippy"` conditions.
To work around
```

> conditions. To work around this, they would either need to allow the lin or have

Reviewer 3: ```suggestion
conditions. To work around this, they would either need to allow the lint or have
```

> this, they would have to add a dummy feature to their `Cargo.toml` in order to
> silence those warnings:

Reviewer: nit:
```suggestion
this, they would either need to allow the lint or have to add a dummy feature
to their `Cargo.toml` in order to silence those warnings:
```
